onnxruntime/core/mlas/lib/logistic.cpp

```diff
@@ -121,6 +121,11 @@ Return Value:
 
         float Value = *Input++;
 
+	// This odd "if test" exists to ensure an input value of NaN carries through
+        // without modification because "std::min" and "std::max" return unreliable results
+        // when NaNs are involved, and it's clear from the test's reference outputs that
+        // they want a NaN on output whenever the input is a NaN.
+        if (Value == Value)
```

Comment:
>if (Value == Value) [](http://example.com/codeflow?start=8&length=19)

this is odd. NaN is sticky. NaN is the expected result for any operatation on NaN, right?

Reply:
Yes, you'd like to think so, but std::min and std::max turn out to be exceptions.

They end up doing a **_comparison_** on the two operands.  
As it happens, if a NaN is involved in either A or B, then (A < B) and (A > B) always (**_ALWAYS_**) return FALSE.  
This makes std::min and std::max return really unfortunate results.
For example, std::min(3.0, NaN) will return NaN, but std::min(NaN, 3.0) will return 3.0.

Basically, one should NOT use std::min or std::max if an input value _might_ be NaN.

Here's a web page which explores this in more depth: 
https://microsoft.public.vc.stl.narkive.com/KmbP2DlQ/std-max-and-nan

So, for both tanh.cpp and logistic.cpp, I just put in that "(Value == Value)" thing, so if Value is a NaN, then the output Value stays NaN (which is the desired result).

Reply:
Thanks for the info. Then I think it will work to adjust the order of the statement below like this:
Value = std::min(std::max(Value, MlasLogisticConstants.LowerRange), MlasLogisticConstants.UpperRange);

Reply:
Thank you for this suggestion.  I will give it more thought tomorrow when my brain is fresh, and maybe code that in and compile it and try it.  Stay tuned!

Reply:
Based on this doc: https://en.cppreference.com/w/cpp/algorithm/max . std::max is implemented as `(a < b) ? b : a; ` , so if a is nan, then it returns nan. similar to min.

We can also replace std::max with `(a < b) ? b : a; ` explicitly, in case std::max is not implemented consistently across different library.

Reply:
I'm always a fan of using simpler calls rather than library calls.
So, I guess we could do this in two steps:
```
   float v_tmp;
   v_tmp = (Value < MlasLogisticConstants.LowerRange) ? MlasLogisticConstants.LowerRange : Value;
   Value = (v_tmp > MlasLogisticConstants.UpperRange) ? MlasLogisticConstants.UpperRange : v_tmp;
```